lib: Improve password fields

skobyda opened the pull request:
- Adds show/hide feature
- Adds a minimum to a progress bar
- Changes width of a progress bar
- Shows message indicating the strength of the password

Fixes https://github.com/cockpit-project/cockpit/issues/19199

[Screencast from 2023-08-22 14-30-21.webm](https://github.com/cockpit-project/cockpit/assets/42733240/231f89ec-04b4-474e-adb1-95b405f5e778)

skobyda commented:
Currently we these 4 levels of messages for passwords "Weak password" < "Good password" < "Strong password" < "Excellent password".
I'm not sure about the wording of the "Good password". It has score more than 33 but less than 66 (and is shown in yellow color). It's a password which is neither good nor bad.
Maybe it should be "acceptable password"? 
WDTY @garrett

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -106,19 +127,35 @@ export const PasswordFormFields = ({
-                </div>
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput className="check-passwords" type={passwordHidden ? "password" : "text"} id={idPrefix + "-pw1"}
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230822-125449-71e7c813-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -106,19 +127,35 @@ export const PasswordFormFields = ({
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordHidden(!passwordHidden)}
+                            aria-label={ passwordHidden ? _("Show password") : _("Hide password")}>
+                            {passwordHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230822-125449-71e7c813-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -131,8 +168,20 @@ export const PasswordFormFields = ({
-                           value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput type={passwordConfirmHidden ? "password" : "text"} id={idPrefix + "-pw2"} autoComplete="new-password"
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230822-125449-71e7c813-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -131,8 +168,20 @@ export const PasswordFormFields = ({
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordConfirmHidden(!passwordConfirmHidden)}
+                            aria-label={passwordConfirmHidden ? _("Show confirmation password") : _("Hide confirmation password")}>
+                            {passwordConfirmHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230822-125449-71e7c813-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt commented:
needsdesign → moving to draft

garrett commented:
![image](https://github.com/cockpit-project/cockpit/assets/10246/05b006ba-9c6a-49f5-adb3-a1f7723fad42)

The contrast between the text and its background is not high enough. The color of the text should be the darker gold.

You can see this in PF demos:

![image](https://github.com/cockpit-project/cockpit/assets/10246/cfcda94e-4dc4-4b70-b08e-41c046205bd2)
https://www.patternfly.org/components/helper-text/html-demos/#helper-text-on-a-form

Here, you can see how PF uses brighter versions for icons, progress bars, etc. and the darker related color for text:
 
![image](https://github.com/cockpit-project/cockpit/assets/10246/c5983e75-b5ea-4858-9316-421278d4c918)
https://www.patternfly.org/components/alert/html/#alert-examples

Colors are specified here:

![image](https://github.com/cockpit-project/cockpit/assets/10246/8adc0352-df94-416c-9853-2353b7983323)
https://www.patternfly.org/design-foundations/colors/#primary-colors

- icon: `--pf-v5-global--warning-color--100`
- text: `--pf-v5-global--warning-color--200`

This isn't just a problem with the warning colors either, but it's most noticeable with it, as it has the least contrast. The other colors need to use the text variants too.

garrett commented:
There's a strange intermediate step where the strength meter is shown in a blue state without text, right at the beginning of typing:

![image](https://github.com/cockpit-project/cockpit/assets/10246/c144a27f-7c04-442a-8a7e-e20a47ce0f69)

garrett commented:
> I'm not sure about the wording of the "Good password". It has score more than 33 but less than 66 (and is shown in yellow color). It's a password which is neither good nor bad.
> Maybe it should be "acceptable password"?

Yes. :+1:

skobyda commented:
Updated: Changed to the text color, got rid of the blue progress bar without any text and from now on using "Acceptable" instead of "Good" password:

![Screenshot from 2023-09-18 14-02-44](https://github.com/cockpit-project/cockpit/assets/42733240/93bd5298-f13e-4e0c-90dc-b997c4fe8771)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -77,20 +82,34 @@ export const PasswordFormFields = ({
                     });
         } else {
-            setPasswordStrength("");
+            setPasswordStrength();
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-011601-1bae20a1-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-052537-1ade659f-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -106,19 +125,35 @@
-                </div>
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput className="check-passwords" type={passwordHidden ? "password" : "text"} id={idPrefix + "-pw1"}
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -106,19 +125,35 @@
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordHidden(!passwordHidden)}
+                            aria-label={ passwordHidden ? _("Show password") : _("Hide password")}>
+                            {passwordHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -131,8 +166,20 @@
-                           value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput type={passwordConfirmHidden ? "password" : "text"} id={idPrefix + "-pw2"} autoComplete="new-password"
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -131,8 +166,20 @@
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordConfirmHidden(!passwordConfirmHidden)}
+                            aria-label={passwordConfirmHidden ? _("Show confirmation password") : _("Hide confirmation password")}>
+                            {passwordConfirmHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20230918-120248-8fd6ef98-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

garrett reviewed (changes requested):
It's a lot better. Thanks!

However, the meter (progress bar) doesn't update based on the strength.

[Screencast from 2023-09-20 09-59-43.webm](https://github.com/cockpit-project/cockpit/assets/10246/1bf48a36-7f82-498c-9386-2d601b34bef3)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -106,19 +125,35 @@ export const PasswordFormFields = ({
-                </div>
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput className="check-passwords" type={passwordHidden ? "password" : "text"} id={idPrefix + "-pw1"}
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -106,19 +125,35 @@ export const PasswordFormFields = ({
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordHidden(!passwordHidden)}
+                            aria-label={ passwordHidden ? _("Show password") : _("Hide password")}>
+                            {passwordHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -131,8 +166,20 @@ export const PasswordFormFields = ({
-                           value={passwordConfirm} onChange={(_event, value) => { setConfirmPassword(value); change("password_confirm", value) }} />
+                <InputGroup>
+                    <InputGroupItem isFill>
+                        <TextInput type={passwordConfirmHidden ? "password" : "text"} id={idPrefix + "-pw2"} autoComplete="new-password"
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -131,8 +166,20 @@ export const PasswordFormFields = ({
+                    <InputGroupItem>
+                        <Button
+                            variant="control"
+                            onClick={() => setPasswordConfirmHidden(!passwordConfirmHidden)}
+                            aria-label={passwordConfirmHidden ? _("Show confirmation password") : _("Hide confirmation password")}>
+                            {passwordConfirmHidden ? <EyeIcon /> : <EyeSlashIcon />}
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231016-125913-03c3b351-fedora-38-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231030-135959-d54860fd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

skobyda commented:
Fixed the issue @garrett 

[Screencast from 2023-10-17 15-06-03.webm](https://github.com/cockpit-project/cockpit/assets/42733240/9c50c190-0faa-4a2c-b1d5-cc4f563aaa8b)

garrett commented on pkg/lib/cockpit-components-password.scss:
```diff
@@ -1,7 +1,12 @@
+@import "global-variables";
+@import "@patternfly/patternfly/utilities/Text/text.scss";
+
 .ct-password-strength-meter {
   grid-gap: var(--pf-v5-global--spacer--xs);
 
   .pf-v5-c-progress__description, .pf-v5-c-progress__status {
     display: none;
   }
+
+  inline-size: var(--pf-v5-global--spacer--2xl);
 }
```
```suggestion
.ct-password-strength-meter {
  grid-gap: var(--pf-v5-global--spacer--xs);
  inline-size: var(--pf-v5-global--spacer--2xl);

  .pf-v5-c-progress__description, .pf-v5-c-progress__status {
    display: none;
  }
}
```

Parent CSS should be grouped together at the top. (This doesn't change anything, but it's better when it's not scattered.)

garrett reviewed (approved):
This was waiting on the password meter fix. Thanks! Approved.

(But please move that line of CSS.)

martinpitt commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -120,19 +122,19 @@ export const PasswordFormFields = ({
                             variant="control"
                             onClick={() => setPasswordHidden(!passwordHidden)}
-                            aria-label={passwordHidden ? _("Show password") : _("Hide password")}>
+                            aria-label={ passwordHidden ? _("Show password") : _("Hide password")}>
```
This is inconsistent spacing, and it also doesn't belong into the "lib: Change width of password progress" commit. This bit should be reverted.

martinpitt commented on pkg/lib/cockpit-components-password.jsx:
```diff
@@ -120,19 +122,19 @@ export const PasswordFormFields = ({
                 </InputGroup>
                 <div>
                     <Progress id={idPrefix + "-meter"}
-                              className={"ct-password-strength-meter " + variant}
```
This previous indentation is better to read IMHO.

martinpitt replied:
Ah, other code in that file uses 4-space indentation as well, so ok.

martinpitt reviewed (changes requested):
Thanks! The code changes look good, but the committed pixel changes [look](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-011601-1bae20a1-fedora-38-storage/log.html) [like](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-011601-1bae20a1-fedora-38-other/log.html) they revert #19520. I'll rebase this PR, fix the above two little issues, and re-do the pixel updates.

martinpitt commented:
That looks much saner -- most tests are okay, just the [accounts tests](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19220-20231031-045839-c175744c-fedora-38-other/log.html) have the updated pixels for the reveal buttons. I'll push these.

martinpitt reviewed (approved):
Thanks!
